File: scheduling/scheduling-annotation-processor/src/main/java/ru/tinkoff/kora/scheduling/annotation/processor/QuartzSchedulingGenerator.java
package ru.tinkoff.kora.scheduling.annotation.processor;

import com.squareup.javapoet.*;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.annotation.processor.common.RecordClassBuilder;

import javax.annotation.processing.Filer;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.io.IOException;
import java.util.Optional;
import java.util.function.Predicate;

public class QuartzSchedulingGenerator {
    private static final ClassName koraQuartzJobClassName = ClassName.get("ru.tinkoff.kora.scheduling.quartz", "KoraQuartzJob");
    private static final ClassName schedulingTelemetryClassName = ClassName.get("ru.tinkoff.kora.scheduling.common.telemetry", "SchedulingTelemetry");
    private static final ClassName schedulingTelemetryFactoryClassName = ClassName.get("ru.tinkoff.kora.scheduling.common.telemetry", "SchedulingTelemetryFactory");
    private static final ClassName triggerClassName = ClassName.get("org.quartz", "Trigger");
    private static final ClassName schedulerClassName = ClassName.get("org.quartz", "Scheduler");
    private static final ClassName triggerBuilderClassName = ClassName.get("org.quartz", "TriggerBuilder");
    private static final ClassName cronScheduleBuilderClassName = ClassName.get("org.quartz", "CronScheduleBuilder");
    private final Types types;
    private final Elements elements;
    private final Filer filer;
    private final TypeMirror scheduleWithTriggerTypeMirror;
    private final TypeMirror scheduleWithCronTypeMirror;

    public QuartzSchedulingGenerator(ProcessingEnvironment processingEnv) {
        this.types = processingEnv.getTypeUtils();
        this.elements = processingEnv.getElementUtils();
        this.filer = processingEnv.getFiler();

        this.scheduleWithTriggerTypeMirror = this.elements.getTypeElement("ru.tinkoff.kora.scheduling.quartz.ScheduleWithTrigger").asType();
        this.scheduleWithCronTypeMirror = this.elements.getTypeElement("ru.tinkoff.kora.scheduling.quartz.ScheduleWithCron").asType();
    }

    public void generate(TypeElement type, ExecutableElement method, TypeSpec.Builder module, SchedulingTrigger trigger) throws IOException {
        var jobClassName = this.generateJobClass(type, method);
        var typeMirror = type.asType();

        var component = MethodSpec.methodBuilder(type.getSimpleName() + "_" + method.getSimpleName() + "_Job")
            .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
            .returns(jobClassName)
            .addParameter(schedulingTelemetryFactoryClassName, "telemetryFactory")
            .addParameter(TypeName.get(typeMirror), "object");


        if (this.types.isSameType(trigger.triggerAnnotation().getAnnotationType(), this.scheduleWithTriggerTypeMirror)) {
            var tag = CommonUtils.<AnnotationMirror>parseAnnotationValue(this.elements, trigger.triggerAnnotation(), "value");
            var triggerParameter = ParameterSpec.builder(triggerClassName, "trigger").addAnnotation(AnnotationSpec.get(tag)).build();
            component.addParameter(triggerParameter);
        } else if (this.types.isSameType(trigger.triggerAnnotation().getAnnotationType(), this.scheduleWithCronTypeMirror)) {
            var identity = Optional.ofNullable(CommonUtils.<String>parseAnnotationValue(elements, trigger.triggerAnnotation(), "identity"))
                .filter(Predicate.not(String::isBlank))
                .orElse(type.getQualifiedName() + "#" + method.getSimpleName());
            var cron = CommonUtils.<String>parseAnnotationValue(elements, trigger.triggerAnnotation(), "value");
            var cronSchedule = CodeBlock.of("$S", cron);
            var configPath = CommonUtils.<String>parseAnnotationValue(elements, trigger.triggerAnnotation(), "config");
            if (configPath != null && !configPath.isBlank()) {
                var configClassName = this.generateCronConfigRecord(type, method, cron);
                var b = MethodSpec.methodBuilder(configClassName.simpleName())
                    .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                    .returns(configClassName)
                    .addParameter(ClassName.get("com.typesafe.config", "Config"), "config");
                if (cron != null && !cron.isBlank()) {
                    b.addCode("if (!config.hasPath($S)) {\n  return new $T($S);\n}\n", configPath, configClassName, cron);
                }
                module.addMethod(b.addCode("""
                        var value = config.getValue($S);
                        var cron = switch (value.valueType()) {
                            case STRING -> value.unwrapped().toString();
                            default -> throw ru.tinkoff.kora.config.common.extractor.ConfigValueExtractionException.unexpectedValueType(value, com.typesafe.config.ConfigValueType.STRING);
                        };
                        """.stripIndent(), configPath)
                    .addCode("return new $T(cron);\n", configClassName)
                    .build());
                component.addParameter(configClassName, "config");
                cronSchedule = CodeBlock.of("config.cron()");
            } else {
                if (cron == null || cron.isBlank()) {
                    throw new ProcessingErrorException("Either value() or config() annotation parameter must be provided", method, trigger.triggerAnnotation());
                }
            }
            component.addCode("""
                var trigger = $T.newTrigger()
                  .withIdentity($S)
                  .withSchedule($T.cronSchedule($L))
                  .build();
                """.stripIndent(), triggerBuilderClassName, identity, cronScheduleBuilderClassName, cronSchedule.toString());
        } else {
            // never gonna happen
            throw new IllegalStateException();
        }
        component.addCode("var telemetry = telemetryFactory.get($T.class, $S);\n", typeMirror, method.getSimpleName().toString());

        component
            .addCode("return new $T(telemetry, object, trigger);\n", jobClassName);

        module.addMethod(component.build());
    }

    private ClassName generateCronConfigRecord(TypeElement type, ExecutableElement method, String defaultCron) throws IOException {
        var configRecordName = CommonUtils.getOuterClassesAsPrefix(method) + "CronConfig";
        var packageName = this.elements.getPackageOf(type).getQualifiedName().toString();

        var rb = new RecordClassBuilder(configRecordName)
            .addModifier(Modifier.PUBLIC);
        if (defaultCron != null && !defaultCron.isBlank()) {
            rb.addComponent("cron", TypeName.get(String.class), CodeBlock.of("$S", defaultCron));
        } else {
            rb.addComponent("cron", TypeName.get(String.class));
        }
        rb.writeTo(this.filer, packageName);

        return ClassName.get(packageName, configRecordName);
    }


    private ClassName generateJobClass(TypeElement type, ExecutableElement method) throws IOException {
        var className = CommonUtils.getOuterClassesAsPrefix(type) + type.getSimpleName() + "_" + method.getSimpleName() + "_Job";
        var packageName = this.elements.getPackageOf(type).getQualifiedName().toString();
        var typeMirror = type.asType();
        var callJob = method.getParameters().isEmpty()
            ? CodeBlock.of("ctx -> object.$L()", method.getSimpleName())
            : CodeBlock.of("object::$L", method.getSimpleName());

        var typeSpec = TypeSpec.classBuilder(className)
            .superclass(koraQuartzJobClassName)
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addField(TypeName.get(typeMirror), "object", Modifier.PRIVATE, Modifier.FINAL)
            .addMethod(MethodSpec.constructorBuilder()
                .addModifiers(Modifier.PUBLIC)
                .addParameter(schedulingTelemetryClassName, "telemetry")
                .addParameter(TypeName.get(typeMirror), "object")
                .addParameter(triggerClassName, "trigger")
                .addCode("super(telemetry, $L, trigger);\n", callJob)
                .addCode("this.object = object;\n")
                .build())
            .build();

        var javaFile = JavaFile.builder(packageName, typeSpec).build();
        javaFile.writeTo(this.filer);

        return ClassName.get(packageName, className);
    }
}

File: scheduling/scheduling-ksp/src/main/kotlin/ru/tinkoff/kora/scheduling/ksp/QuartzSchedulingGenerator.kt
package ru.tinkoff.kora.scheduling.ksp

import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.symbol.KSAnnotation
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.writeTo
import ru.tinkoff.kora.common.Tag
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.getOuterClassesAsPrefix
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.writeTagValue

class QuartzSchedulingGenerator(val env: SymbolProcessorEnvironment) {
    private val koraQuartzJobClassName: ClassName = ClassName("ru.tinkoff.kora.scheduling.quartz", "KoraQuartzJob")
    private val schedulingTelemetryClassName: ClassName = ClassName("ru.tinkoff.kora.scheduling.common.telemetry", "SchedulingTelemetry")
    private val schedulingTelemetryFactoryClassName: ClassName = ClassName("ru.tinkoff.kora.scheduling.common.telemetry", "SchedulingTelemetryFactory")
    private val triggerClassName: ClassName = ClassName("org.quartz", "Trigger")
    private val schedulerClassName: ClassName = ClassName("org.quartz", "Scheduler")
    private val triggerBuilderClassName: ClassName = ClassName("org.quartz", "TriggerBuilder")
    private val cronScheduleBuilderClassName: ClassName = ClassName("org.quartz", "CronScheduleBuilder")


    fun generate(type: KSClassDeclaration, function: KSFunctionDeclaration, builder: TypeSpec.Builder, trigger: SchedulingTrigger) {
        val jobClassName = generateJobClass(type, function)
        val typeClassName = type.toClassName()
        val component = FunSpec.builder("_" + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_Job")
            .returns(jobClassName)
            .addParameter("telemetryFactory", schedulingTelemetryFactoryClassName)
            .addParameter("target", typeClassName);

        when (trigger.annotation.shortName.getShortName()) {
            "ScheduleWithTrigger" -> {
                val tag = trigger.annotation.findValue<KSAnnotation>("value")!!
                val tagAnnotationSpec = AnnotationSpec.builder(Tag::class)
                    .addMember(tag.findValue<List<KSType>>("value")!!.writeTagValue("value"))
                    .build()

                val triggerParameter = ParameterSpec.builder("trigger", triggerClassName)
                    .addAnnotation(tagAnnotationSpec)
                    .build()
                component.addParameter(triggerParameter)
            }

            "ScheduleWithCron" -> {
                val identity = trigger.annotation.findValue<String>("identity").let {
                    if (it.isNullOrBlank()) {
                        type.qualifiedName!!.asString() + "#" + function.simpleName.getShortName()
                    } else {
                        it
                    }
                }
                val cron = trigger.annotation.findValue<String>("value") ?: ""
                var cronSchedule = CodeBlock.of("%S", cron)
                val configPath = trigger.annotation.findValue<String>("config")
                if (!configPath.isNullOrBlank()) {
                    val configClassName = this.generateCronConfigRecord(type, function, cron)
                    val b = FunSpec.builder(configClassName.simpleName)
                        .returns(configClassName)
                        .addParameter("config", ClassName("com.typesafe.config", "Config"))
                    if (cron.isNotBlank()) {
                        b.addCode("if (!config.hasPath(%S)) {\n  return %T(%S);\n}\n", configPath, configClassName, cron)
                    }
                    b.addCode(
                        """
                            val value = config.getValue(%S)
                            if (value.valueType() != com.typesafe.config.ConfigValueType.STRING) {
                              throw ru.tinkoff.kora.config.common.extractor.ConfigValueExtractionException.unexpectedValueType(value, com.typesafe.config.ConfigValueType.STRING)
                            }
                            return %T(value.unwrapped().toString());
                        """.trimIndent(), configPath, configClassName
                    )
                    builder.addFunction(b.build())
                    component.addParameter("config", configClassName);
                    cronSchedule = CodeBlock.of("config.cron")
                }
                component.addCode(
                    """
                val trigger = %T.newTrigger()
                  .withIdentity(%S)
                  .withSchedule(%T.cronSchedule(%L))
                  .build();
                """.trimIndent() + "\n", triggerBuilderClassName, identity, cronScheduleBuilderClassName, cronSchedule.toString()
                );
            }
        }
        component.addCode("val telemetry = telemetryFactory.get(%T::class.java, %S);\n", typeClassName, function.simpleName.getShortName())

        component
            .addCode("return %T(telemetry, target, trigger);\n", jobClassName)

        builder.addFunction(component.build())
    }


    private fun generateJobClass(type: KSClassDeclaration, method: KSFunctionDeclaration): ClassName {
        val className: String = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + method.simpleName.getShortName() + "_Job"
        val packageName: String = type.packageName.asString()
        val callJob = if (method.parameters.none { !it.hasDefault }) {
            CodeBlock.of("{ ctx -> target.%L() }", method.simpleName.getShortName())
        } else {
            CodeBlock.of("{ ctx -> target.%L(ctx) }", method.simpleName.getShortName())
        }
        val typeClassName = type.toClassName()
        val typeSpec: TypeSpec = TypeSpec.classBuilder(className)
            .superclass(koraQuartzJobClassName)
            .addSuperclassConstructorParameter(CodeBlock.of("telemetry"))
            .addSuperclassConstructorParameter(callJob)
            .addSuperclassConstructorParameter(CodeBlock.of("trigger"))
            .addProperty(
                PropertySpec.builder("target", typeClassName, KModifier.PRIVATE, KModifier.FINAL)
                    .initializer("target")
                    .build()
            )
            .primaryConstructor(
                FunSpec.constructorBuilder()
                    .addParameter("telemetry", schedulingTelemetryClassName)
                    .addParameter("target", typeClassName)
                    .addParameter("trigger", triggerClassName)
                    .build()
            )
            .build()
        FileSpec.get(packageName, typeSpec).writeTo(env.codeGenerator, false, listOf(type.containingFile!!))
        return ClassName(packageName, className)
    }

    private fun generateCronConfigRecord(type: KSClassDeclaration, function: KSFunctionDeclaration, defaultCron: String): ClassName {
        val configClassName = type.getOuterClassesAsPrefix() + type.simpleName.getShortName() + "_" + function.simpleName.getShortName() + "_CronConfig"
        val constructor = FunSpec.constructorBuilder()
        if (defaultCron.isBlank()) {
            constructor.addParameter("cron", STRING)
        } else {
            constructor.addParameter(
                ParameterSpec.builder("cron", STRING)
                    .defaultValue(CodeBlock.of("%S", defaultCron))
                    .build()
            )
        }
        val configType = TypeSpec.classBuilder(configClassName)
            .addModifiers(KModifier.PUBLIC, KModifier.DATA)
            .addProperty(
                PropertySpec.builder("cron", STRING)
                    .initializer("cron")
                    .build()
            )
            .primaryConstructor(constructor.build())
            .build()
        FileSpec.get(type.packageName.asString(), configType).writeTo(env.codeGenerator, false, listOf(type.containingFile!!))
        return ClassName(type.packageName.asString(), configClassName)
    }
}
